Control/Parallel/Strategies.hs

```diff
@@ -432,19 +432,33 @@ rpar  x = case (par# x) of { _ -> Done x }
 #endif
 {-# INLINE rpar  #-}
 
--- | instead of saying @rpar `dot` strat@, you can say
--- @rparWith strat@.  Compared to 'rpar', 'rparWith'
+-- | Perform a computation in parallel using a strategy.
 --
---  * does not exit the `Eval` monad
---
---  * does not have a built-in `rseq`, so for example `rparWith r0`
---    behaves as you might expect (it is a strategy that creates a
---    spark that does no evaluation).
+-- @
+-- rparWith strat x
+-- @
 --
+-- will spark a thread to perform @strat x@ in parallel. Note
+-- that @rparWith strat@ is /not/ the same as @rpar `dot` strat@.
+-- Specifically, @rpar `dot` strat@ always produces a value in
+-- WHNF, while @rparWith strat@ need not.
 --
+-- > rparWith r0 = r0
+-- > rparWith rpar = rpar
+-- > rparWith rseq = rpar
 rparWith :: Strategy a -> Strategy a
 #if __GLASGOW_HASKELL__ >= 702
-rparWith s = rpar `dot` s
+-- The intermediate `Lift` box is necessary, in order to avoid a built-in
+-- `rseq` in `rparWith`. In particular, we want rparWith r0 = r0, not
+-- rparWith r0 = rpar.
+rparWith s a = do
+  l <- rpar r
```

Comment:
This creates (nearly) always a fizzled spark, because `l` is immediately pattern matched. A lazy pattern match is needed.

Reply:
The pattern match is lazy (it's in the argument to `return`). Try your test!

Reply:
Moreover, if you inline everything manually (including the `runRW#`) you'll end up with the same thing as the old implementation. We just shouldn't do that because inlining the `realWorld#` is like asking the optimizer to completely mangle our intentions.

Reply:
(FYI, search the `bytestring` source code for `accursedUnutterablePerformIO` if you want a more detailed explanation.)

Reply:
Sorry I was wrong with my comment.